lib/SIL/TypeLowering.cpp

```diff
@@ -2150,10 +2164,10 @@ TypeConverter::getLoweredLocalCaptures(AnyFunctionRef fn) {
   // add it as the final capture. Otherwise, add a fake hidden capture for
   // the dynamic 'Self' metatype.
   if (selfCapture.hasValue()) {
-    captures.insert(*selfCapture);
+    captures.push_back(*selfCapture);
```

Comment:
Is it possible for `self` to have been captured explicitly and thus already be in the list?

Reply:
Ah, yes, that looks like a separate logic bug, line 2134 here: if it's already capturing `self` and there's another occurrence it does fall through to be added to the list. Will fix to `mergeFlags` there as well.

Reply:
I meant that `self` could be in the `localCaptures` list, right? Say, for an explicit capture with `[weak self]`.

Reply:
Oh, yes, in which case it'd be assigned to `selfCapture` if it has the dynamic self type, or it would be a "normal" capture if not. Either way, now all occurrences get collapsed into one and it would be added to the transitive capture list only once.

And then (the only reason for tracking `selfCapture` separately, I think) if it needs dynamic self but it wasn't captured explicitly, the `else if` fake hidden capture of the meta type.

Reply:
But nothing checks that when going through the normal captures, no?

Maybe I just don't understand this code. @slavapestov, @jckarter?

Reply:
'selfCapture' is a special thing because of how DynamicSelfType works. When you reify the runtime metadata for 'Self', we load the isa pointer from the 'self' value, eg

```
func myGeneric<T>(_: T) -> T.Type {
  return T.self
}

class C {
  func returnsSelf() -> Self {
    _ = { print(myGeneric(self)) }() // we're calling myGeneric() with T := DynamicSelfType
    return self
  }
}
```

Here, the closure captures `self` strong, so it can recover the generic metadata for passing to `myGeneric()` from the isa pointer of `self`. However it's possible to refer to the dynamic `Self` type implicitly inside a closure, for example:

```
class C {
  required init() {}

  func returnsSelf() -> Self {
    return { .init() }()
  }
}
```

The type of the closure here is inferred as `() -> Self`. The closure has no captures, but we're calling `Self.init()`, so the question becomes how do you recover the dynamic type? This is why we have the special `selfCapture` value -- we just add an implicit metatype capture to the closure.

Now I think `selfCapture` should not have capture flags; the value is always a metatype which cannot have ownership qualifiers. Can you add assertions to that effect?

Reply:
If this is really about capturing `Self` and not `self` then it definitely needs a rename. But I don't think it's just about that, since it's also inferred from the VarDecl `self`. Or maybe the variable's being reused but shouldn't be.

Reply:
I added this code in https://github.com/apple/swift/commit/2a59ad7d0561f865be1b8d9525526359ddc6cff5, but it was two years ago. If I recall correctly it was just about `Self`. The reason `self` comes into play is if you're capturing `self` and its not weak, you don't have to also capture `Self` :) But this was a long time ago.